draft: introducing gatsby themes

ChristopherBiscardi opened the pull request:
Introductory blog post for initial theming support

signalwerk commented:
Hey @ChristopherBiscardi 
That looks nice. Plus I guess it's related to https://github.com/gatsbyjs/gatsby/issues/2662 but I'm not sure what are the steps to get it running. I see your repo, but how do I use it now in my site? I need to setup a whole gatsby and then i can reference the theme? I was always hoping gatsby will at one point be that simple, that I can just have a repo with a package.json and my content of the page and the rest will be handled by the theme in my separate npm repo. A bit like [mdx-deck](https://github.com/jxnblk/mdx-deck) is doing it.

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+add support for our favorite styling technique, transform markdown,
+and just about anything else you can imagine.
+
+At it's core, a gatsby site is a combination of functionality centered
```
```suggestion
At its core, a Gatsby site is a combination of functionality centered
```

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+and just about anything else you can imagine.
+
+At it's core, a gatsby site is a combination of functionality centered
+around a single config file, gatsby-config.js. This config file
```
```suggestion
around a single config file, `gatsby-config.js`. This config file
```

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+At it's core, a gatsby site is a combination of functionality centered
+around a single config file, gatsby-config.js. This config file
+controls an assortment of site metadata, data type mapping, and most
+importantly plugins. Plugins contain large amounts of customizable
```
```suggestion
importantly, plugins. Plugins contain large amounts of customizable
```

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+
+# Scaling Gatsby
+
+Creating one Gatsby site is great but we had to start from scratch and
```
Maybe re-phrase this? Something like:

> Creating a single Gatsby site works super well. The power of `gatsby-config.js`, plugins, and more coalesce to make the experience a breeze. However, what if we want to re-use this configuration on our _next_ site? Sure, we _could_ clone a boilerplate each time, but that gets old, quickly. Wouldn't it be great if we could re-use our `gatsby-config.js` across projects? Introducing... starters!

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+use starters. Starters are basically whole Gatsby sites that can be
+scaffolded through the gatsby cli. This helps you start your project
+by cloning the boilerplate, installing dependencies, and clearing Git
+history. The community around gatsby has built a lot of different
```
```suggestion
history. The community around Gatsby has built a lot of different
```

should always be capitalized! I'll stop fixing these though, you get the idea :)

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+upstream. There's another approach to boilerplate that has become
+popular in recent years that fixes some problems with the boilerplate
+approach such as updating with upstream. One such project is
+create-react-app. In the gatsby world, we can improve on starters
```
Maybe explain this further? _Why_ is create-react-app so novel? (It's because it internalizes configuration and we get any updates, changes, etc. for _free_ merely by bumping the version)

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+
+# Why Themes?
+
+Defining themes as the composition of Gatsby sites allows us to start
```
```suggestion
Defining themes as the base composition unit of Gatsby sites allows us to start
```

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+Defining themes as the composition of Gatsby sites allows us to start
+solving a variety of use cases. For example, when a site gets built as
+part of a wider product offering it's often the case that one team
+will build out a set of functionality, including branding elements,
```
```suggestion
will build out a suite of functionality, including branding elements,
```

DSchau replied:
(this is more of an opinion here!)

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+# Mechanics of Theming
+
+At a base level, theming simply combines the gatsby-config of the
+theme with the gatsby-config of your site. Since it's an experimental
```
Worth discussing a plan/roadmap here. Maybe an RFC if we don't already have one?

jlengstorf replied:
```suggestion
theme with the `gatsby-config.js` of your site. Since it's an experimental
```

DSchau commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
````diff
@@ -0,0 +1,154 @@
+Then in our theme's createPage call, we simply use the helper to let
+the user optionally override the default component.
+
+```
````
```suggestion
```title:gatsby-node.js
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+At a base level, theming simply combines the gatsby-config of the
+theme with the gatsby-config of your site. Since it's an experimental
+feature, we use an experimental namespace to declare themes in
+gatsby-config.
```
```suggestion
the config.
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
````diff
@@ -0,0 +1,154 @@
+```
+
+Themes also function as plugins and any config passed into the theme
+in your gatsby-config will also be passed to your theme's gatsby-\*
````
```suggestion
in your `gatsby-config.js` will also be passed to your theme's `gatsby-*.js`
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+in your gatsby-config will also be passed to your theme's gatsby-\*
+files as plugin options. This allows themes to override any settings
+inherited from the theme's own plugin declarations or apply gatsby
+lifecycle hooks such as on-create-page.
```
```suggestion
lifecycle hooks such as [`onCreatePage`](https://www.gatsbyjs.org/docs/node-apis/#onCreatePage).
```

ChristopherBiscardi replied:
shouldn't this be `(/docs/node-apis/#onCreatePage)`?

DSchau replied:
@ChristopherBiscardi yes 👍

jlengstorf replied:
Fixed, sorry.

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+inherited from the theme's own plugin declarations or apply gatsby
+lifecycle hooks such as on-create-page.
+
+# Next Steps
```
```suggestion
## Next Steps
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+
+# Next Steps
+
+## Sub Themes and Overriding
```
```suggestion
### Sub Themes and Overriding
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+functionality for turning markdown into pages, processing components
+into documentation, and even processing images.
+
+# Scaling Gatsby
```
```suggestion
## Scaling Gatsby
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+we have the base for themes. We can encode portions of our
+gatsby-config as themes and re-use them across sites.
+
+# Why Themes?
```
```suggestion
## Why Themes?
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+the customization of various branding elements through our
+gatsby-config.
+
+# Mechanics of Theming
```
```suggestion
## Mechanics of Theming
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+This is just the first step and it enables us to experiment with
+further improvements in userland before merging them into
+core. Sub-theming, for example, is a critical part of a theming
+ecosystem that is currently missing from core. Overriding theme
```
```suggestion
ecosystem that is currently missing from Gatsby. Overriding theme
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+core. Sub-theming, for example, is a critical part of a theming
+ecosystem that is currently missing from core. Overriding theme
+elements is possible on a coarse level right now in userland. If, for
+example, a theme defines a set of pages using createPage we can define
```
```suggestion
example, a theme defines a set of pages using [`createPage`](https://www.gatsbyjs.org/docs/actions/#createPage) we can define
```

jlengstorf replied:
@ChristopherBiscardi I edited this to not use the domain. That's my fault, sorry.

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
````diff
@@ -0,0 +1,154 @@
+}
+```
+
+Then in our theme's createPage call, we simply use the helper to let
````
```suggestion
Then in our theme's `createPage` call, we simply use the helper to let
```

jlengstorf commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+
+This doesn't allow us to make more granular overrides of different
+components, but it does allow us to replace the rendering of pages and
+other whole elements.
```
Is the intention of this to set up something that will be fixed soon? It's not entirely clear if this section is saying, "You should do this," or, "This is the only way to do things _now_, but we're going to fix it shortly."

jlengstorf replied:
Oh, that's nice! Tag me on that PR, please, because I'd love to dig through it. I have some initial thoughts that I'll keep to myself to not start bikeshedding in a PR about something else. 😄

ChristopherBiscardi commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+## Themes
+
+If a single gatsby-config encodes the functionality of a whole Gatsby
+site then if we can compose the gatsby-config data structure together
```
nice link

ChristopherBiscardi commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+
+# Mechanics of Theming
+
+At a base level, theming simply combines the gatsby-config of the
```
ugh, thanks for removing "simply".

ChristopherBiscardi commented:
> I see your repo, but how do I use it now in my site? I need to setup a whole gatsby and then i can reference the theme?

In the repo, you'll see a theme and two usages of that theme. The theme is the https://github.com/gatsbyjs/gatsby-starter-blog with a couple modifications. The theme is not published to npm yet, but I plan to publish it under an experimental namespace before publishing this post so that people who read this post can immediately try it out.

You'll be most interested in the [minimal example](https://github.com/ChristopherBiscardi/gatsby-theme-examples/tree/4f36ea53bc0d4254c84a7b2834d6598dadcd8892/ui-surfaces/blog-b-minimal), which relies on the theme for pretty much everything except the icon and provides it's own content.

> I was always hoping gatsby will at one point be that simple, that I can just have a repo with a package.json and my content of the page and the rest will be handled by the theme in my separate npm repo.

That is how this will work. If you look at the example repo you'll notice it's a multi-package repo using yarn workspaces, which if you're not familiar means that the theme operates exactly as an npm package would. As we start using themes more we'll discover that some plugins don't expect to be used from inside an npm package due to the way they handle paths, but these problems are pretty trivial and will be smoothed over by a couple PRs as they're discovered.

> A bit like [mdx-deck](https://github.com/jxnblk/mdx-deck) is doing it.

Could you be more specific about what you like about mdx-deck's model? My understanding of mdx-deck is that since it has a hard dependency on a specific styling approach, it can use those primitives. Gatsby themes operate at a slightly higher level of abstraction right now where this styling approach can be defined in the theme itself. The official Gatsby themes I'm working on will use [Emotion](https://emotion.sh/) and allow you to pass in theme variables in the theme config, as mentioned in this blog post.

ChristopherBiscardi commented:
I applied most of the suggested comments. 

@jlengstorf Should links include the domain? or be from the root like `/docs/blog/` The [contribution doc](https://www.gatsbyjs.org/docs/how-to-contribute/#contributing-to-the-blog) shows the latter, but some of the suggestions here are the former so it might be out of date?

DSchau commented:
@ChristopherBiscardi taking another look at this now.

> Should links include the domain?

Nope! Should be relative links, so sorry for the confusion there!

signalwerk commented:
sorry @ChristopherBiscardi I didn't see your response to my questions tow days ago and answered [here](https://github.com/gatsbyjs/gatsby/issues/2662) and had some more questions while using it [here](https://github.com/ChristopherBiscardi/gatsby-theme-examples/issues). I'm not sure where to find a reference, since your blog is the only reference I could found, I thought it would probably help to have a example that answers most of the questions from newbies like me ;-)

LekoArts commented:
Could you clarify here/in the post on how to convert a Gatsby Starter to a theme? As you might know I‘ve got multiple starters which e.g. all use emotion/styled component theming. How would I change the starter to make it usable as a starter and theme?

ChristopherBiscardi commented:
@LekoArts The main changes to make are

* create an empty index.js in the root of the starter. 
  - This is because themes are also run as plugins and plugins require the empty index.js
* Adjust any paths for plugins so that they point where you want them to. This can be a path into to the theme, to the site, a path that will work for both depending on if you're running the site or just the theme by itself, or anything else you want. 

Look at the example [in the themes repo](https://github.com/ChristopherBiscardi/gatsby-theme-examples/blob/master/themes/gatsby-theme-blog/gatsby-config.js) which is a port of the [gatsby-starter-blog](https://github.com/gatsbyjs/gatsby-starter-blog).

I'll see about adding a section to the blog with this information (or possible a different blog post depending on how much information is necessary).

@signalwerk no worries :) The questions are useful when crafting content like this blog post. I addressed the issues you filed, etc

LekoArts commented:
But the starter would still work standalone with these changes? As I don’t want to manage two repos for one starter.

LekoArts commented:
> ah, you want it to work as a starter and a theme?

Exactly.

> I'm curious as to the use case there, but yes. 

I'm not sure about that right now but that's something we have to see if there's a usecase

shannonbux commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,154 @@
+starters for various use cases including blogging, working with
+material design, and documentation.
+
+The problem with starters is that they're one-offs. Starters are
```
This is where it starts to get interesting for me. I think the stuff before this could be condensed; do readers need a definition of starters?

jlengstorf replied:
We could maybe just link to starters for those who aren't sure what this means to avoid distracting from the point of the post?

ChristopherBiscardi replied:
While we're mostly reaching out to people who already know what themes are and why they'd want them with this blog post, I included this section to point out how starters are one step up on "the reusability chain": more reuse than a regular site but less than themes. "How are starters and themes different" is also a question I see as going to come up a lot, so defining starters next to themes helps people contrast the approaches side-by-side.

shannonbux replied:
Ok, I can see the purpose then! Cool!

signalwerk commented:
@shannonbux I like your perspective and how you see it through the readers eyes!

LekoArts commented:
Why did you create a new community on Spectrum instead of requesting a new channel on the Gatsby community?
On a related note: The Spectrum community is kinda handled shabby (so I can understand) but I'd try to use the existing Gatsby community. @gatsbyjs/inkteam should probably get ownership of that community and create a more welcoming community there (+ your themes channel).

ChristopherBiscardi commented:
@shannonbux 

> I prefer using "you" instead of "we"

Happy to adhere to Gatsby's norms, made this change. 

> What is the ultimate end goal of this blogpost? 

I want people to head to the Spectrum community where we're colocating conversation about themes and the theming ecosystem.

> Do you have an example of a theme you could describe? Like one that you wish you had or are excited about building?

There's a blog theme that was built by doing a conversion of the gatsby-starter-blog repo. It's in the mentioned examples repo in the blogpost.

ChristopherBiscardi commented:
@LekoArts The new spectrum chat is specifically for themes and is intended to be temporary while the feature and ecosystem develops. Since I'm leading the feature and will be dealing with most of the topics filed there for the near future, a separate community was deemed to be the right approach. We'll merge it back into the official channel in the future. This was something I discussed with @KyleAMathews and @calcsam (so tagging them both in case they have different feelings about it).

ChristopherBiscardi commented:
@shannonbux @jlengstorf @DSchau I'd love to get some review with the recent changes.

shannonbux commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,171 @@
+other whole elements. Component Shadowing, a more granular method of
+overriding, is already in the works.
+
+If you want to be involved in the development of theming for Gatsby,
```
Nice ending!

shannonbux commented on docs/blog/2018-10-26-introducing-gatsby-themes/index.md:
```diff
@@ -0,0 +1,171 @@
+---
+
+Gatsby is a powerful platform for building marketing sites, blogs,
+ecommerce frontends, and more. We can source data from static files
```
Just noticing a couple more "we"'s that confused me. Another chance for "you", if you ask me ;)

ChristopherBiscardi replied:
heh, habits die hard.

ChristopherBiscardi replied:
@shannonbux handled.

ChristopherBiscardi commented:
@shannonbux I've removed the rest of the "we"s
